pkg/networkserver/mac/adr.go

```diff
@@ -145,10 +145,21 @@ func txPowerStep(phy *band.Band, from, to uint8) float32 {
 }
 
 func AdaptDataRate(ctx context.Context, dev *ttnpb.EndDevice, phy *band.Band, defaults ttnpb.MACSettings) error {
-	if len(dev.RecentADRUplinks) == 0 {
+	if dev.MACState == nil || len(dev.MACState.RecentUplinks) == 0 {
 		return nil
 	}
 
+	adrUplinks := dev.MACState.RecentUplinks
+	for i := len(dev.MACState.RecentUplinks) - 1; i >= 0; i-- {
+		up := dev.MACState.RecentUplinks[i]
+		if (up.Payload.MType == ttnpb.MType_UNCONFIRMED_UP || up.Payload.MType == ttnpb.MType_CONFIRMED_UP) &&
+			up.Payload.GetMACPayload().FCnt >= dev.MACState.LastADRChangeFCntUp {
```

Comment:
Shouldn't this be comparing full FCnt? Assuming that the payload is not being changed.